cfgmgr/buffer_check_headroom_mellanox.lua

```diff
@@ -81,41 +82,49 @@ local function get_number_of_pgs(keyname)
-if new_pg ~= nil then
-    if get_number_of_pgs(new_pg) ~= 0 then
-        no_input_pg = false
-        new_pg = 'BUFFER_PG_TABLE:' .. new_pg
-    end
+-- Fetch all the PGs in APPL_DB, and store them into a hash table
+local pg_keys = redis.call('KEYS', 'BUFFER_PG_TABLE:' .. port .. ':*')
+local all_pgs = {}
+for i = 1, #pg_keys do
+    local profile = redis.call('HGET', pg_keys[i], 'profile')
+    all_pgs[pg_keys[i]] = profile
+end
+
+-- Fetch all the pending PGs, and store them into the hash table
+-- Overwrite any existing entries
+local pending_pg_keys = redis.call('KEYS', '_BUFFER_PG_TABLE:' .. port .. ':*')
+for i = 1, #pending_pg_keys do
+    local profile = redis.call('HGET', pending_pg_keys[i], 'profile')
+    -- Remove the leading underscore when storing it into the hash table
+    all_pgs[string.sub(pending_pg_keys[i], 2, -1)] = profile
+    table.insert(debuginfo, 'debug:pending entry: ' .. pending_pg_keys[i] .. ':' .. profile)
+end
+
+if new_pg ~= nil and get_number_of_pgs(new_pg) ~= 0 then
+    all_pgs['BUFFER_PG_TABLE:' .. new_pg] = input_profile_name
 end
 
--- Fetch all the PGs, accumulate the sizes
+-- Handle all the PGs, accumulate the sizes
 -- Assume there is only one lossless profile configured among all PGs on each port
 table.insert(debuginfo, 'debug:other overhead:' .. accumulative_size)
-local pg_keys = redis.call('KEYS', 'BUFFER_PG_TABLE:' .. port .. ':*')
-for i = 1, #pg_keys do
-    local profile = redis.call('HGET', pg_keys[i], 'profile')
+for pg_key, profile in pairs(all_pgs) do
     local current_profile_size
-    if profile ~= 'ingress_lossy_profile' and (no_input_pg or new_pg ~= pg_keys[i]) then
-        if profile ~= input_profile_name and not no_input_pg then
-            local referenced_profile = redis.call('HGETALL', 'BUFFER_PROFILE_TABLE:' .. profile)
-            for j = 1, #referenced_profile, 2 do
-                if referenced_profile[j] == 'size' then
-                    current_profile_size = tonumber(referenced_profile[j+1])
-                end
-            end
-        else
-            current_profile_size = input_profile_size
-            profile = input_profile_name
+    if profile ~= input_profile_name then
+        local referenced_profile_size = redis.call('HGET', 'BUFFER_PROFILE_TABLE:' .. profile, 'size')
+        if not referenced_profile_size then
+            referenced_profile_size = redis.call('HGET', '_BUFFER_PROFILE_TABLE:' .. profile, 'size')
+            table.insert(debuginfo, 'debug:pending profile: ' .. profile)
         end
-        accumulative_size = accumulative_size + current_profile_size * get_number_of_pgs(pg_keys[i])
-        table.insert(debuginfo, 'debug:' .. pg_keys[i] .. ':' .. profile .. ':' .. current_profile_size .. ':' .. get_number_of_pgs(pg_keys[i]) .. ':accu:' .. accumulative_size)
+        current_profile_size = tonumber(referenced_profile_size)
+    else
+        current_profile_size = input_profile_size
+        profile = input_profile_name
```

Comment:
Why do we need this assignment? We reach L119 when profile is equal to 'input_profile_name'

Reply:
Nice catch. We do not need this. Removed.